packages/purpose-process/src/services/purposeService.ts

```diff
@@ -203,6 +209,60 @@ export function purposeServiceBuilder(
       });
       await repository.createEvent(event);
     },
+    async rejectPurposeVersion({
+      purposeId,
+      versionId,
+      rejectionReason,
+      organizationId,
+      correlationId,
+    }: {
+      purposeId: PurposeId;
+      versionId: PurposeVersionId;
+      rejectionReason: string;
+      organizationId: TenantId;
+      correlationId: string;
+    }): Promise<void> {
+      logger.info(`Rejecting Version ${versionId} in Purpose ${purposeId}`);
+
+      const purpose = await retrievePurpose(purposeId, readModelService);
+      const eservice = await retrieveEService(
+        purpose.data.eserviceId,
+        readModelService
+      );
+      if (organizationId !== eservice.producerId) {
+        throw organizationIsNotTheProducer(organizationId);
+      }
+
+      const purposeVersion = retrievePurposeVersion(versionId, purpose);
+
+      if (purposeVersion.state !== purposeVersionState.waitingForApproval) {
+        throw notValidVersionState(purposeVersion.id, purposeVersion.state);
+      }
+
+      if (!rejectionReason) {
+        throw missingRejectionReason();
+      }
```

Comment:
You can move this at the start, in order to not retrieve anything if the reason is missing

Reply:
Updated